services/src/contexts/postgres.rs

```diff
@@ -0,0 +1,2597 @@
+        };
+
+        Ok(PostgresContext {
+            default_session_id: session.id(),
+            task_manager: Default::default(),
+            thread_pool: create_rayon_thread_pool(0),
+            exe_ctx_tiling_spec,
+            query_ctx_chunk_size,
+            pool,
+            volumes: Default::default(),
+        })
+    }
+
+    // TODO: check if the datasets exist already and don't output warnings when skipping them
+    #[allow(clippy::too_many_arguments)]
+    pub async fn new_with_data(
+        config: Config,
+        tls: Tls,
+        dataset_defs_path: PathBuf,
+        provider_defs_path: PathBuf,
+        layer_defs_path: PathBuf,
+        layer_collection_defs_path: PathBuf,
+        exe_ctx_tiling_spec: TilingSpecification,
+        query_ctx_chunk_size: ChunkByteSize,
+    ) -> Result<Self> {
+        let pg_mgr = PostgresConnectionManager::new(config, tls);
+
+        let pool = Pool::builder().build(pg_mgr).await?;
+        let created_schema = Self::create_schema(pool.get().await?).await?;
+
+        let session = if created_schema {
+            let session = SimpleSession::default();
+            Self::create_default_session(pool.get().await?, session.id()).await?;
+            session
+        } else {
+            Self::load_default_session(pool.get().await?).await?
+        };
+
+        let app_ctx = PostgresContext {
+            default_session_id: session.id(),
+            task_manager: Default::default(),
+            thread_pool: create_rayon_thread_pool(0),
+            exe_ctx_tiling_spec,
+            query_ctx_chunk_size,
+            pool,
+            volumes: Default::default(),
+        };
+
+        if created_schema {
+            info!("Populating database with initial data...");
+
+            let ctx = app_ctx.session_context(session);
+
+            let mut db = ctx.db();
+            add_layers_from_directory(&mut db, layer_defs_path).await;
+            add_layer_collections_from_directory(&mut db, layer_collection_defs_path).await;
+
+            add_datasets_from_directory(&mut db, dataset_defs_path).await;
+
+            add_providers_from_directory(&mut db, provider_defs_path, &[]).await;
+        }
+
+        Ok(app_ctx)
+    }
+
+    async fn is_schema_initialized(
+        conn: &PooledConnection<'_, PostgresConnectionManager<Tls>>,
+    ) -> Result<bool> {
+        let stmt = match conn.prepare("SELECT TRUE from geoengine;").await {
+            Ok(stmt) => stmt,
+            Err(e) => {
+                if let Some(code) = e.code() {
+                    if *code == SqlState::UNDEFINED_TABLE {
+                        info!("Initializing schema.");
+                        return Ok(false);
+                    }
+                }
+                return Err(error::Error::TokioPostgres { source: e });
+            }
+        };
+
+        let _row = conn.query(&stmt, &[]).await?;
+
+        Ok(true)
+    }
+
+    #[allow(clippy::too_many_lines)]
```

Comment:
We could put the create statements into a .sql file (which gets proper highlighting and formatting in VSCode) and then include the contents into this files via the `include_str!` macro.

Reply:
the text is not valid SQL though, because we are replacing some stuff in placeholders.

Reply:
Can we get rid of the placeholders?

The guys from sqlx said that it is much better if you don't allow placeholders since changing the inputs makes your created schema invalid/different.

It is only two insert statements. We could make them separately?

Reply:
7a90b48f1737a64e40a3ab6af6d16deaa12da3ef